src/codegen/compile.ml

```diff
@@ -3583,14 +3724,15 @@ module Arr = struct
       (* No need to check the lower bound, we interpret idx as unsigned *)
       (* Check the upper bound *)
       get_idx ^^
-      get_array ^^ Heap.load_field len_field ^^
+      get_array ^^ len env ^^
       G.i (Compare (Wasm.Values.I32 I32Op.LtU)) ^^
       E.else_trap_with env "Array index out of bounds" ^^
 
       get_idx ^^
       compile_add_const header_size ^^
       compile_mul_const element_size ^^
       get_array ^^
+      Tagged.load_forwarding_pointer env ^^
```

Comment:
Indeed, this could offer nice optimization potential considering that array element accesses are frequent. I will do this and measure the performance gain.

Reply:
I tried the change below to avoid multiple forwarding resolution. Interestingly, it is slower with around 0.5% performance degrade for array-intense GC benchmark cases...

```
      let set_array = G.setter_for get_array in
      get_array ^^ Tagged.load_forwarding_pointer env ^^ set_array ^^
      (* No need to check the lower bound, we interpret idx as unsigned *)
      (* Check the upper bound *)
      get_idx ^^
      get_array ^^ Heap.load_field len_field ^^
      G.i (Compare (Wasm.Values.I32 I32Op.LtU)) ^^
      E.else_trap_with env "Array index out of bounds" ^^

      get_idx ^^
      compile_add_const header_size ^^
      compile_mul_const element_size ^^
      get_array ^^
      G.i (Binary (Wasm.Values.I32 I32Op.Add))
      ```

Reply:
Joachim just opened a PR to remove the bounds check anyway #3853, which would remove the redundant forwarding anyway.